llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -542,6 +550,91 @@ class MachineSchedWAWEdges : public WAWEdges {
 class SWPWAWEdges : public WAWEdges {
   void apply(ScheduleDAGInstrs *DAG) override { WAWEdges::apply(DAG); }
 };
+
+class WAWStickyRegistersEdges : public ScheduleDAGMutation {
+  void apply(ScheduleDAGInstrs *DAG) override {
+    MachineFunction &MF = DAG->MF;
+    const TargetRegisterInfo *TRI = MF.getSubtarget().getRegisterInfo();
+    auto *RI = static_cast<const AIEBaseRegisterInfo *>(TRI);
+    const MachineBasicBlock *MBB = DAG->getBB();
+
+    if (MBB) {
+      BitVector AllRegs(RI->getNumRegs());
+      AllRegs.reset();
+      // Here, we analyze which sticky registers are explicitly redefined
+      // of read. We also track all instructions implicitly reading or
+      // defining such registers.
+      std::map<const Register, SmallVector<const MachineInstr *, 16>> RegMIsMap;
+      for (const MachineInstr &MI : *MBB) {
+        for (const MachineOperand &MOP : MI.operands()) {
+          if (!MOP.isReg())
+            continue;
+
+          const Register Reg = MOP.getReg();
+          if (!Reg.isPhysical() || !RI->isReservedStickyReg(Reg))
+            continue;
+
+          if ((!MOP.isImplicit() && (MOP.isDef() || MOP.readsReg())) ||
+              (MOP.isImplicit() && MOP.readsReg())) {
+            AllRegs.set(Reg);
+
+          } else if (MOP.isImplicit() && MOP.isDef()) {
+            // Instruction that could have a dependency removal.
+            // We track it because of the next heuristic.
+            RegMIsMap[Reg].push_back(&MI);
+          }
+        }
+      }
+
+      auto IsLoad = [&](const MachineInstr *MI) -> bool {
+        return MI->mayLoad();
+      };
+      auto IsStore = [&](const MachineInstr *MI) -> bool {
+        return MI->mayStore();
+      };
+
+      // This is the heuristic component. We catch basically cases where
+      // registers are only defined by loads or store within a block,
+      // For example, cases like exemplified below (MBB):
+      //   [sequence non-defining sticky regs. instructions.]
+      //   VST.CONV ...
+      //   VST.CONV ...
+      //   VST.CONV ...
+      //   VST.CONV ...
+      // In this case, by removing dependencies between pairs of VST.CONVs,
+      // we give too much freedom to the scheduler to do good, but also
+      // not good choices. In this way, we filter those cases off.
+      for (auto RMIs : RegMIsMap) {
+        const Register Reg = RMIs.first;
+        SmallVector<const MachineInstr *, 16> &MIs = RMIs.second;
+        // The first thing to test is the tuning parameter: we only consider
+        // cases where the number of memory ops are <= the threshold.
+        if (MIs.size() <= WAWStickyRegistersMemOpsThreshold &&
+            ((all_of(MIs, IsLoad) || all_of(MIs, IsStore))))
```

Comment:
Hmm. I guess you only need to remember NumLoads and NumStores to evaluate this condition

Reply:
As I am interested on cases where all  instructions must be loads or stores (per sticky register),  I will need a per-register set of counters. I think in this way the code a bit more clear.